### PR TITLE
NEXT-3743 - Fix inadequate entity definition resolving in Elasticsearch filter parsing that show when filtering for a translated field of a related entity is executed in Shopware 6.5

### PR DESCRIPTION
**Pull request by SpiGAndromeda**
### 1. Why is this change necessary?

If a filter is applied on an Elasticsearch query that uses a translated field of an entity that is related to the product entity, a `RuntimeException` is thrown in the `getTranslatedField`method of the `EntityDefinitionQueryHelper`.

The filter parsing uses the `getField` method of the `EntityDefinitionQueryHelper` to get the field even if the field name contains relations. The same is not done for the definition itself. As the root definition is passed to the value parsing this would mostly lead to a null return of the `getField` method, which is unexpected behavoir. However, if the field is `TranslatedField, the `getTranslatedField` method is called. This method only searches for the field in the fields of the passed translation (without deeper resolving). The field won't be found an an execption is thrown indicating that the property is missing.

### 2. What does this change do, exactly?

This change introduces a private `getDefinition` method in the `CriteriaParser` to resolve the definition over it's relations to a definition that actually contains the field. This method is called in the `parseEqualsAnyFilter` method, the `parseEqualsFilter` method and the `parseRangeFilter` method. The original definition will still be used for the query creation. If the field does not exist in the resolved entity, an exception is still thrown.

### 3. Describe each step to reproduce the issue or behaviour.

The bug only occurs on translated fields of entities that are related to the product entity. One example is the `shortCode` field in the unit entity. The exception will be thrown if a filter (equals, equals any or range) is applied to this field.

### 4. Please link to the relevant issues (if any).

https://github.com/shopware/shopware/issues/3743

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

### 6. Related PR

There is closed (an merged) PR for this changes implemented in SW 6.6. This is a backport of this changes to SW 6.5. https://github.com/shopware/shopware/pull/5739

### CODE REVIEW

File: changelog/_unreleased/2024-12-04-fixed-elasticsearch-filter-parsing-of-translated-fields-in-product-related-entities.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+---
+title: Fixed Elasticsearch Filter parsing of translated fields in product-related entities
+issue: NEXT-37804
+author: Martin Bens
+author_email: martin.bens@it-bens.de
+author_github: @spigandromeda
+---
+# Core
+
+* Backported the `getAssociatedDefinition` method of the `EntityDefinitionQueryHelper` from NEXT-34674
```
Comment (vienthuong) [inline changelog/_unreleased/2024-12-04-fixed-elasticsearch-filter-parsing-of-translated-fields-in-product-related-entities.md]: Added method ...

We don't use the keyword `Backported`

**Comment by vienthuong** [comment]
Hi @SpiGAndromeda 
Thanks again for your contribution
And sorry for not being clear about the tests, in our pipeline, the major flag (v6.6.0.0) is activated/deactivated on different CI jobs so you don't need to add your own logic for it.

The only thing you need to adjust is to make sure the test is passed when the feature flag is on and off. So for example if your response is different you can do this

```
if (Feature::isActive('v6.6.0.0') {
    static::assertEquals([ ... ], $response);
} else {
    static::assertEquals([ ... ], $response);
}
```

If the response are the same in both environment, you can leave it as it is.

**Comment by SpiGAndromeda** [comment]
Changed that and overwrote the commit.

**Comment by vienthuong** [comment]
Hi @SpiGAndromeda 

Your changes has been merged into 6.5.x branch 🎉 , it will be released within the next 6.5.x patch release

**Pull request closed without merging**